HADOOP-14217. Support colon in Hadoop Path

yufeldman opened the pull request.

steveloughran commented:
I think some new Abstract FS Contract test will be needed to see how filesystems actually handle ":" chars; having the path support it is just one part of the problem

yufeldman commented:
will do

yufeldman commented:
I have added few tests to FileSystemContractBaseTest and updated HDFS and WebHDFS to not run those tests, as HDFS and subsequently WebHdfs currently do not support colon in any portion of the path including FileName(s).
I did test with RawLocalFileSystem, S3(a,n). Since I can't test Azure, Swift, other FSs would be great to get feedback whether they support colon and can run those tests.

vrozov commented:
It will be good to change Path constructors that take `String child` as an argument to use `new Path(null, null, child)` instead of `new Path(child)`. It should help to handle cases like `new Path("file:/", "abc:/def");`.

yufeldman commented:
@vrozov I don't think we can change existing constructors, only add one and one can use at their discretion, but we can't change all the code that is already written and uses existing constructors, especially the code that we use in our applications implicitly (two, three, ..., times removed from what we are trying to do)

vrozov commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/Path.java:
```diff
@@ -214,6 +213,12 @@ public Path(String pathString) throws IllegalArgumentException {
     // uri path is the rest of the string -- query & fragment not supported
     String path = pathString.substring(start, pathString.length());
 
+    // add "./" in front of Linux relative paths so that a path containing
```
Try to avoid code duplication by introducing `initialize(scheme, authority, path)` that both `Path(String)` and `Path(String, String, String)` may use.

vrozov replied:
@yufeldman Also, it is not clear why is this change necessary? The comment is misleading as `scheme` is already initialized.

yufeldman replied:
Imagine string passed to that ctor is: `a:b` so with new changes `a:` is not really a scheme.
Now when you try to construct URI with null scheme and authority URI will interpret `a:` as scheme unless path is started with `/`

vrozov replied:
Should there be an explicit check for `null` scheme? Also, how ":/abc" is handled? Should it be considered as "abc" under ":", instead of an empty scheme that is not valid?

yufeldman replied:
Please check URI ctor for the behaviors you mentioned. Precautions I am taking is to avoid exceptions in URI ctor.
Also I suggest you look at the doc I attached to JIRA (below is a paragraph from the doc):
There are couple of cases that will not be handled by this approach or behavior will change.
Namely:
1. URI that consists only of the scheme - e.g. “x:” will result in Path: “./x:” (Today trying to
create a URI from this string will result in error: “Relative path in absolute URI”)
2. Path that has colon as last character before “/” - e.g. “abc:/bcd”. This will be interpreted
as scheme “abc” and path “/bcd” - the same as today
3. String such as: “urn:example:animal:ferret:nose” will be interpreted as
“./urn:example:animal:ferret:nose” (Today trying to create a URI from this string will
result in error: “Relative path in absolute URI”)

vrozov replied:
I refer to a check for `scheme == null` prior to making path relative on Linux (`path = "./ + path`).
According to your specification (item 2), ":/abc" will be interpreted as "/abc" with an empty scheme and still throw an exception. Will it be better to handle it as "./:/abc" path?

vrozov commented:
@yufeldman My suggestion is not to change all the code that is already written. I refer to `Path(String parent, String child)` constructor on line 114 that calls `new Path(child)`. There is no need to parse `child`, only `parent` may have scheme and authority.

yufeldman commented:
@vrozov  I agree in principal that in Hadoop we do not consider that "child" can have scheme and it would help handling cases with ":/" in "child portion of path, but  according to URI resolve method it can be different configurations of merging parent and child.
We would need very extensive testing to cover those cases. Also I believe there is method in Path:
`mergePaths` that people are using to signify what you were saying.
I would like to here more opinions on changing behavior of `Path(Path, Path)` to make it behave more like `mergePaths` method

vrozov commented:
@yufeldman I am not sure that URI `resolve()` method (RFC2396 5.2) is directly applicable here even though `Path` uses `URI.resolve()` in its implementation. `Path` constructor refers to `parent` and `child`, that is different from constructing absolute URI from base and current, where current may be an absolute URI. For `Path(parent, child)`, I would not expect `child` to be an absolute Path.

yufeldman commented:
@vrozov I don't disagree with you. I would like to get consensus (more input from stakeholders) on changing this behavior. This JIRA is least invasive.

yufeldman commented:
@vrozov how about having a hangout meeting - maybe it's simpler that way?

vrozov commented:
@yufeldman sure, if that will be easier.

yufeldman commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/Path.java:
```diff
@@ -196,8 +196,7 @@ public Path(String pathString) throws IllegalArgumentException {
     int slash = pathString.indexOf('/');
-    if ((colon != -1) &&
-        ((slash == -1) || (colon < slash))) {     // has a scheme
+    if ((colon != -1) && (colon == (slash-1))) {     // has a scheme
```
@vrozov You have a point - if I change behavior here to colon > 0 it will support parts of path that are just ":".
The only thing is different behavior when you merge two paths - `URI.resolve()` resolves it differently
E.g:
`/abc/:` and `:/:/` versus `/abc/:` and `/:/:/` 
In first case result will be `/abc/:/:/:/` and in the second case `/abc/:/:/`

vrozov replied:
@yufeldman In the first case, you merge absolute and relative path, in the second case, there is an attempt to merge two absolute path, should they be handled differently if merging two absolute path should be allowed at all?

vrozov commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/Path.java:
```diff
@@ -214,6 +213,12 @@ public Path(String pathString) throws IllegalArgumentException {
 
+    // add "./" in front of Linux relative paths so that a path containing
+    // a colon e.q. "a:b" will not be interpreted as scheme "a".
+    if (!WINDOWS && !path.isEmpty() && path.charAt(0) != '/') {
```
Will it be better to change the condition to `if (!WINDOWS && scheme == null && colon != -1)`? I also have a concern regarding `WINDOWS`, as ":" is file system dependent, not OS specific, but this concern is beyond this PR and JIRA.

yufeldman replied:
What if it is: `/abc:bcd`. scheme == null, colon != -1? I don't think we want to add `/` once more. It will be than: `.//abc:bcd`

vrozov replied:
`(scheme == null && slash > colon)`? What should happen in case of "./abc" and "./abc:"?

yufeldman replied:
sorry, I don't understand what you are implying. slash can be after colon - there could be many slashes and many colons.

vrozov replied:
I refer to the first slash and the first colon in the passed `pathString` parameter:
```
int colon = pathString.indexOf(':');
int slash = pathString.indexOf('/');
```
My understanding is that prepending "./" to the path, ensures that "/" is before ":" to avoid interpreting character sequence that `Path` does not recognized as scheme to be interpreted as scheme by `URI`. If there is no colon or slash is already before a colon, there is no need to add "./".

yufeldman replied:
Yes we have more info to make the check.  What is the advantage of this change?
Three params ctor does have this check in place as well.

vrozov replied:
To avoid String concatenation and creation of unnecessary String objects. Prior to the patch, there was no "./" in case of a long path and it can (should?) be avoided with the patch if that long path does not contain ":".
I am not a committer for Hadoop and it is a recommendation to consider only.

elek commented:
Can one of the admins verify this patch?

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 112 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 4 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 27 | Maven dependency ordering for branch |
| +1 | mvninstall | 1415 | trunk passed |
| +1 | compile | 1437 | trunk passed |
| +1 | checkstyle | 189 | trunk passed |
| +1 | mvnsite | 221 | trunk passed |
| +1 | shadedclient | 1336 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 146 | trunk passed |
| 0 | spotbugs | 191 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 346 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 25 | Maven dependency ordering for patch |
| +1 | mvninstall | 117 | the patch passed |
| +1 | compile | 1037 | the patch passed |
| +1 | javac | 1037 | the patch passed |
| -0 | checkstyle | 153 | root: The patch generated 7 new + 126 unchanged - 0 fixed = 133 total (was 126) |
| +1 | mvnsite | 167 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 749 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 129 | the patch passed |
| +1 | findbugs | 341 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 586 | hadoop-common in the patch passed. |
| -1 | unit | 2652 | hadoop-hdfs in the patch failed. |
| -1 | asflicense | 63 | The patch generated 15 ASF License warnings. |
| | | 11121 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.TestReadStripedFileWithDecodingDeletedData |
|   | hadoop.hdfs.TestFileChecksum |
|   | hadoop.hdfs.TestApplyingStoragePolicy |
|   | hadoop.hdfs.TestErasureCodingExerciseAPIs |
|   | hadoop.hdfs.TestEncryptionZonesWithKMS |
|   | hadoop.hdfs.TestPread |
|   | hadoop.hdfs.TestDFSStripedOutputStreamWithRandomECPolicy |
|   | hadoop.hdfs.TestFetchImage |
|   | hadoop.hdfs.tools.TestDFSZKFailoverController |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=18.09.7 Server=18.09.7 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-269/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/269 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 65d1a967b4b4 4.15.0-52-generic #56-Ubuntu SMP Tue Jun 4 22:49:08 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 7f1b76c |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/1/artifact/out/diff-checkstyle-root.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/1/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/1/testReport/ |
| asflicense | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/1/artifact/out/patch-asflicense-problems.txt |
| Max. process+thread count | 2440 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 44 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 4 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 21 | Maven dependency ordering for branch |
| +1 | mvninstall | 1074 | trunk passed |
| +1 | compile | 1062 | trunk passed |
| +1 | checkstyle | 148 | trunk passed |
| +1 | mvnsite | 141 | trunk passed |
| +1 | shadedclient | 947 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 116 | trunk passed |
| 0 | spotbugs | 126 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| -1 | findbugs | 164 | hadoop-hdfs in trunk failed. |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 26 | Maven dependency ordering for patch |
| +1 | mvninstall | 112 | the patch passed |
| +1 | compile | 1048 | the patch passed |
| +1 | javac | 1048 | the patch passed |
| -0 | checkstyle | 129 | root: The patch generated 7 new + 126 unchanged - 0 fixed = 133 total (was 126) |
| +1 | mvnsite | 136 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 636 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 116 | the patch passed |
| +1 | findbugs | 296 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 550 | hadoop-common in the patch passed. |
| -1 | unit | 5023 | hadoop-hdfs in the patch failed. |
| +1 | asflicense | 62 | The patch does not generate ASF License warnings. |
| | | 11952 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.TestErasureCodingPoliciesWithRandomECPolicy |
|   | hadoop.hdfs.TestErasureCodingMultipleRacks |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-269/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/269 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 1000a3817efd 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 62efb63 |
| Default Java | 1.8.0_212 |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/2/artifact/out/branch-findbugs-hadoop-hdfs-project_hadoop-hdfs.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/2/artifact/out/diff-checkstyle-root.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/2/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/2/testReport/ |
| Max. process+thread count | 4053 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 109 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 4 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 24 | Maven dependency ordering for branch |
| +1 | mvninstall | 1262 | trunk passed |
| +1 | compile | 1164 | trunk passed |
| +1 | checkstyle | 159 | trunk passed |
| +1 | mvnsite | 173 | trunk passed |
| +1 | shadedclient | 1137 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 127 | trunk passed |
| 0 | spotbugs | 147 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| -1 | findbugs | 204 | hadoop-hdfs in trunk failed. |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 25 | Maven dependency ordering for patch |
| +1 | mvninstall | 151 | the patch passed |
| +1 | compile | 1107 | the patch passed |
| +1 | javac | 1107 | the patch passed |
| -0 | checkstyle | 151 | root: The patch generated 7 new + 126 unchanged - 0 fixed = 133 total (was 126) |
| +1 | mvnsite | 161 | the patch passed |
| +1 | whitespace | 1 | The patch has no whitespace issues. |
| +1 | shadedclient | 765 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 145 | the patch passed |
| +1 | findbugs | 350 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 608 | hadoop-common in the patch passed. |
| -1 | unit | 6518 | hadoop-hdfs in the patch failed. |
| +1 | asflicense | 64 | The patch does not generate ASF License warnings. |
| | | 14491 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.namenode.TestNameNodeMXBean |
|   | hadoop.hdfs.server.datanode.TestDataNodeUUID |
|   | hadoop.hdfs.server.datanode.TestLargeBlockReport |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-269/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/269 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 8d2bceff5deb 4.15.0-52-generic #56-Ubuntu SMP Tue Jun 4 22:49:08 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / e872ceb |
| Default Java | 1.8.0_222 |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/3/artifact/out/branch-findbugs-hadoop-hdfs-project_hadoop-hdfs.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/3/artifact/out/diff-checkstyle-root.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/3/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/3/testReport/ |
| Max. process+thread count | 2850 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/3/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 42 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 4 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 22 | Maven dependency ordering for branch |
| +1 | mvninstall | 1070 | trunk passed |
| +1 | compile | 1143 | trunk passed |
| +1 | checkstyle | 141 | trunk passed |
| +1 | mvnsite | 162 | trunk passed |
| +1 | shadedclient | 963 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 113 | trunk passed |
| 0 | spotbugs | 113 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| -1 | findbugs | 123 | hadoop-hdfs in trunk failed. |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 25 | Maven dependency ordering for patch |
| +1 | mvninstall | 110 | the patch passed |
| +1 | compile | 1114 | the patch passed |
| +1 | javac | 1114 | the patch passed |
| -0 | checkstyle | 136 | root: The patch generated 7 new + 126 unchanged - 0 fixed = 133 total (was 126) |
| +1 | mvnsite | 135 | the patch passed |
| -1 | whitespace | 0 | The patch has 2 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply |
| +1 | shadedclient | 620 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 109 | the patch passed |
| +1 | findbugs | 292 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 514 | hadoop-common in the patch failed. |
| -1 | unit | 5953 | hadoop-hdfs in the patch failed. |
| -1 | asflicense | 61 | The patch generated 1 ASF License warnings. |
| | | 12930 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.util.TestDiskCheckerWithDiskIo |
|   | hadoop.hdfs.TestDatanodeDeath |
|   | hadoop.hdfs.TestFileCreation |
|   | hadoop.hdfs.TestDFSShell |
|   | hadoop.hdfs.TestErasureCodingExerciseAPIs |
|   | hadoop.hdfs.TestMaintenanceState |
|   | hadoop.hdfs.server.datanode.TestLargeBlockReport |
|   | hadoop.hdfs.qjournal.TestMiniJournalCluster |
|   | hadoop.hdfs.TestHDFSFileSystemContract |
|   | hadoop.hdfs.tools.offlineImageViewer.TestOfflineImageViewerForContentSummary |
|   | hadoop.hdfs.TestErasureCodingPolicies |
|   | hadoop.hdfs.tools.TestECAdmin |
|   | hadoop.hdfs.TestDecommissionWithStriped |
|   | hadoop.hdfs.TestSafeModeWithStripedFile |
|   | hadoop.hdfs.TestReconstructStripedFileWithRandomECPolicy |
|   | hadoop.hdfs.TestByteBufferPread |
|   | hadoop.hdfs.TestDFSStripedOutputStream |
|   | hadoop.hdfs.TestWriteReadStripedFile |
|   | hadoop.hdfs.TestReservedRawPaths |
|   | hadoop.hdfs.TestUnsetAndChangeDirectoryEcPolicy |
|   | hadoop.hdfs.server.blockmanagement.TestUnderReplicatedBlocks |
|   | hadoop.hdfs.server.namenode.ha.TestDFSUpgradeWithHA |
|   | hadoop.hdfs.TestDistributedFileSystem |
|   | hadoop.hdfs.TestErasureCodingPoliciesWithRandomECPolicy |
|   | hadoop.hdfs.TestFileChecksumCompositeCrc |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-269/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/269 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 3d7a4cf7710f 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 00b5a27 |
| Default Java | 1.8.0_212 |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/4/artifact/out/branch-findbugs-hadoop-hdfs-project_hadoop-hdfs.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/4/artifact/out/diff-checkstyle-root.txt |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/4/artifact/out/whitespace-eol.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/4/artifact/out/patch-unit-hadoop-common-project_hadoop-common.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/4/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/4/testReport/ |
| asflicense | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/4/artifact/out/patch-asflicense-problems.txt |
| Max. process+thread count | 4142 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/4/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 46 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 4 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 24 | Maven dependency ordering for branch |
| +1 | mvninstall | 1080 | trunk passed |
| +1 | compile | 1047 | trunk passed |
| +1 | checkstyle | 130 | trunk passed |
| +1 | mvnsite | 132 | trunk passed |
| +1 | shadedclient | 945 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 116 | trunk passed |
| 0 | spotbugs | 159 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 268 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 22 | Maven dependency ordering for patch |
| +1 | mvninstall | 104 | the patch passed |
| +1 | compile | 1007 | the patch passed |
| +1 | javac | 1007 | the patch passed |
| -0 | checkstyle | 141 | root: The patch generated 7 new + 126 unchanged - 0 fixed = 133 total (was 126) |
| +1 | mvnsite | 135 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 658 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 114 | the patch passed |
| +1 | findbugs | 301 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 503 | hadoop-common in the patch failed. |
| -1 | unit | 4885 | hadoop-hdfs in the patch failed. |
| +1 | asflicense | 48 | The patch does not generate ASF License warnings. |
| | | 11716 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ha.TestZKFailoverController |
|   | hadoop.hdfs.server.datanode.TestDirectoryScanner |
|   | hadoop.hdfs.qjournal.server.TestJournalNodeSync |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-269/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/269 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux b1d0d8ce5a92 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 8943e13 |
| Default Java | 1.8.0_222 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/5/artifact/out/diff-checkstyle-root.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/5/artifact/out/patch-unit-hadoop-common-project_hadoop-common.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/5/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/5/testReport/ |
| Max. process+thread count | 4274 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/5/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 43 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 4 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 21 | Maven dependency ordering for branch |
| +1 | mvninstall | 1072 | trunk passed |
| +1 | compile | 1015 | trunk passed |
| +1 | checkstyle | 129 | trunk passed |
| +1 | mvnsite | 131 | trunk passed |
| +1 | shadedclient | 973 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 113 | trunk passed |
| 0 | spotbugs | 170 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 281 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 22 | Maven dependency ordering for patch |
| +1 | mvninstall | 109 | the patch passed |
| +1 | compile | 1042 | the patch passed |
| +1 | javac | 1042 | the patch passed |
| -0 | checkstyle | 136 | root: The patch generated 7 new + 126 unchanged - 0 fixed = 133 total (was 126) |
| +1 | mvnsite | 141 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 649 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 121 | the patch passed |
| +1 | findbugs | 296 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 545 | hadoop-common in the patch passed. |
| -1 | unit | 4813 | hadoop-hdfs in the patch failed. |
| +1 | asflicense | 48 | The patch does not generate ASF License warnings. |
| | | 11689 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.namenode.TestNameNodeMetadataConsistency |
|   | hadoop.hdfs.server.balancer.TestBalancerWithSaslDataTransfer |
|   | hadoop.hdfs.tools.TestDFSZKFailoverController |
|   | hadoop.hdfs.server.balancer.TestBalancer |
|   | hadoop.hdfs.server.namenode.TestFsck |
|   | hadoop.hdfs.server.balancer.TestBalancerRPCDelay |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-269/6/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/269 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 0eb8da722e92 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 4cb22cd |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/6/artifact/out/diff-checkstyle-root.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/6/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/6/testReport/ |
| Max. process+thread count | 4472 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/6/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 35 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 4 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 23 | Maven dependency ordering for branch |
| +1 | mvninstall | 1184 | trunk passed |
| +1 | compile | 1082 | trunk passed |
| +1 | checkstyle | 163 | trunk passed |
| +1 | mvnsite | 160 | trunk passed |
| +1 | shadedclient | 1131 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 141 | trunk passed |
| 0 | spotbugs | 199 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 340 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 24 | Maven dependency ordering for patch |
| +1 | mvninstall | 113 | the patch passed |
| +1 | compile | 1096 | the patch passed |
| +1 | javac | 1096 | the patch passed |
| -0 | checkstyle | 151 | root: The patch generated 7 new + 126 unchanged - 0 fixed = 133 total (was 126) |
| +1 | mvnsite | 173 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 738 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 123 | the patch passed |
| +1 | findbugs | 341 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 565 | hadoop-common in the patch passed. |
| -1 | unit | 75 | hadoop-hdfs in the patch failed. |
| +1 | asflicense | 49 | The patch does not generate ASF License warnings. |
| | | 7636 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-269/7/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/269 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 39113425507a 4.15.0-54-generic #58-Ubuntu SMP Mon Jun 24 10:55:24 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 63c295e |
| Default Java | 1.8.0_222 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/7/artifact/out/diff-checkstyle-root.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/7/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/7/testReport/ |
| Max. process+thread count | 1492 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/7/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 92 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 4 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 25 | Maven dependency ordering for branch |
| +1 | mvninstall | 1184 | trunk passed |
| +1 | compile | 1042 | trunk passed |
| +1 | checkstyle | 153 | trunk passed |
| +1 | mvnsite | 154 | trunk passed |
| +1 | shadedclient | 1117 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 125 | trunk passed |
| 0 | spotbugs | 177 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 293 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 21 | Maven dependency ordering for patch |
| +1 | mvninstall | 108 | the patch passed |
| +1 | compile | 998 | the patch passed |
| +1 | javac | 998 | the patch passed |
| -0 | checkstyle | 155 | root: The patch generated 7 new + 126 unchanged - 0 fixed = 133 total (was 126) |
| +1 | mvnsite | 152 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 739 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 125 | the patch passed |
| +1 | findbugs | 308 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 571 | hadoop-common in the patch passed. |
| -1 | unit | 5834 | hadoop-hdfs in the patch failed. |
| +1 | asflicense | 53 | The patch does not generate ASF License warnings. |
| | | 13213 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.tools.TestDFSZKFailoverController |
|   | hadoop.hdfs.TestRollingUpgrade |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-269/8/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/269 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux bfbd2ce45d3d 4.15.0-54-generic #58-Ubuntu SMP Mon Jun 24 10:55:24 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 3329257 |
| Default Java | 1.8.0_222 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/8/artifact/out/diff-checkstyle-root.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/8/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/8/testReport/ |
| Max. process+thread count | 3095 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-269/8/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.
